File: machine learning models/Kmeans Up-Down Stock Movement Clustering/Python/kmstock.py
import numpy as np
import csv
import random
import os

def km(X, K, max_iters):
    m, n = X.shape
    centroids = np.random.rand(K, n)
    idx = np.zeros((m, 1))
    idx = idx.astype(int)
    old_idx = np.zeros((m, 1))
    old_idx = old_idx.astype(int)

    for j in range(max_iters):
        change = False
        for i in range(m):
            idx[i] = int(np.argmin(np.sum((np.tile(X[i, :], (K, 1)) - centroids)**2, axis = 1)))
            if (idx[i] != old_idx[i]) and (not change):
                change = True
        for k in range(K):
            k = np.array(k)
            index = np.squeeze(idx == k)
            centroids[k,:] = np.mean(np.squeeze(X[np.where(index), :]), axis = 0)

        if not change:
            break

        old_idx = idx

    return idx

cwd = os.path.dirname(__file__)
with open(r"" + cwd + "/sp500_short_period.csv") as csvfile:
    csvData = csv.reader(csvfile)
    datList = []
    for row in csvData:
        datList.append(row)

symbols = np.array(datList.pop(0))

data = np.array(datList)
data = data.astype(np.float)
c = np.double((data[1:np.size(data, 0), :] - data[0:np.size(data, 0) - 1, :]) > 0)
movement = np.transpose(c)

K = 10                          # 10 sectors so that 10 classes
max_iters = 1000                # maximum iterations
random.seed(1234)
idx = km(movement, K, max_iters)

for k in range(K):
    print('\nStocks in group %d moving up together\n' % (k+1))
    k = np.array(k)
    index = np.squeeze(idx == k)
    print(symbols[np.where(index)])

File: machine learning models/Kmeans Up-Down Stock Movement Clustering/MATLAB/kmstock.m
function kmstock()
    clear all; close all; clc;
    [data,symbols,raw] = xlsread('sp500_short_period.xlsx','sp500');
    movement = double((data(2:end,:)-data(1:end-1,:))>0)';
    [m,n] = size(movement);

    K = 10; % 10 sectors
    max_iters = 100;

    [idx,centroids] = km(movement,K,max_iters);

    for i=1:K
        fprintf('\nStocks in group %d moving up together\n',i);
        char(symbols(idx == i))
    end


    function [idx,centroids] = km(X,K,max_iters)
        [m n] = size(X);
        centroids = rand(K,n);
        idx = zeros(m,1);
        old_idx = zeros(m,1);

        for j=1:max_iters
            change = false;
            for i=1:m
                [minval,idx(i)] = min(sum((repmat(X(i,:),K,1)-centroids).^2,2));
                if ((idx(i)~=old_idx(i)) && (change == false))
                    change = true;
                end
            end
            for i=1:K
               centroids(i,:) = mean(X(idx==i,:),1); 
            end

            if (change == false)  
               break;
            end

            old_idx = idx;
        end
    end
end

